stonesoup/types/orbitalstate.py

```diff
@@ -0,0 +1,840 @@
+            :math:`\dot{r}_x, \dot{r}_y, \dot{r}_z` are the
+            corresponding velocity coordinates.
+
+        coordinates = "Keplerian" (Keplarian elements),
+
+            .. math::
+
+                X_{t_0} = [e, a, i, \Omega, \omega, \theta]^{T}
+
+            where :math:`e` is the orbital eccentricity (unitless),
+            :math:`a` the semi-major axis ([length]), :math:`i` the
+            inclination (radian), :math:`\Omega` is the longitude of the
+            ascending node (radian), :math:`\omega` the argument of
+            periapsis (radian), and :math:`\theta` the true anomaly
+            (radian).
+
+        coordinates = "TLE" (Two-Line elements, [2]),
+
+            .. math::
+
+                X_{t_0} = [i, \Omega, e, \omega, M_0, n]^{T}
+
+            where :math:`i` the inclination (radians), :math:`\Omega`
+            is the longitude of the ascending node (radian), :math:`e`
+            is the orbital eccentricity (unitless), :math:`\omega` the
+            argument of perigee (radian), :math:`M_0` the mean anomaly
+            (radian) :math:`n` the mean motion (radian
+            [time] :math:`^{-1}`)
+
+        coordinates = "Equinoctial" (equinoctial elements, [3])
+
+            .. math::
+
+                X_{t_0} = [a, h, k, p, q, \lambda]^{T}
+
+            where :math:`a` the semi-major axis ([length]),
+            :math:`h` is the horizontal component of the
+            eccentricity, :math:`k` is the vertical component of the
+            eccentricity, :math:`p` is the horizontal component of the
+            inclination, :math:`q` is the vertical component of the
+            inclination and :math:`\lambda` is the mean longitude
+
+    The gravitational parameter :math:`\mu = GM` can be defined. If
+    left undefined it defaults to that of the Earth, :math:`3.986004418
+    \, (\pm \, 0.000000008) \times 10^{14} \mathrm{m}^3 \mathrm{s}^{−2}`
+
+    References
+    ----------
+    1. Curtis, H.D. 2010, Orbital Mechanics for Engineering
+    Students (3rd Ed), Elsevier Aerospace Engineering Series
+
+    2. https://spaceflight.nasa.gov/realdata/sightings/SSapplications/\
+    Post/JavaSSOP/SSOP_Help/tle_def.html
+
+    3. Broucke, R. A. & Cefola, P. J. 1972, Celestial Mechanics, Volume
+    5, Issue 3, pp. 303-310
+
+    """
+
+    coordinates = Property(
+        str, default='Cartesian',
+        doc="The parameterisation used on initiation. Acceptable values "
+            "are 'Cartesian' (default), 'Keplerian', 'TLE', or 'Equinoctial'. "
+            "All other inputs will return errors."
+    )
+
+    grav_parameter = Property(
+        float, default=3.986004418e14,
+        doc=r"Standard gravitational parameter :math:`\mu = G M`. The default "
+            r"is :math:`3.986004418 \times 10^{14} \,` "
+            r":math:`\mathrm{m}^3 \mathrm{s}^{-2}`.")
+
+    metadata = Property(
+        dict, default={}, doc="Dictionary containing metadata about orbit"
+    )
+
```

Comment:
It may be worth adding a `__new__` method to return the specific types below, and then that means their `__init__` methods are used to validate arguments (and can also do coordinate conversion?).

Can see something similar in [`stonesoup.types.hypothesis`](https://github.com/dstl/Stone-Soup/blob/7707806d79a00b270934eeeb3ad5d11e83ac1281/stonesoup/types/hypothesis.py#L122-L130)

Reply:
'May' or 'must'?

Also, don't understand. Do you mean, use `__new__` to return a some general `KeplerianOrbitalState`, `TLEOrbitalState` , etc (below), then specifically `__init__` one of those? How would that work?

Reply:
Just a suggestion. You can use `__new__` to decide what type to return, which in turn will then create an instance of that class instead (which will also call it's `__init__` method, etc.)

Would be something like:
```python
def __new__(cls, ...):
    if coordinates == "TLE":
        return super().__new__(TLEOrbitalState)
    elif ...:
        ...
    else:
        raise TypeError("Unknown coordinate system...")
```

Reply:
Unless there's a strong argument for doing this then I'll defer. It's going to require me learning new stuff, and likely moving significant code around, which will introduce a fair delay.

Reply:
I can make this work such that `a = OrbitalState(x, coordinates='TLE',...)` initiates a  `TLEOrbitalState`, but now initialisation via `b = TLEOrbitalState(x,..)` chucks an error ("no property _property_state_vector"), though `c = TLEOrbitalState(x, coordinates='TLE',...)` still works, despite the redundant keyword.

However:

I've come to the conclusion that `KeplerianOrbitalState`, `TLEOrbitalState`, etc classes are a bad idea. They are fundamentally the same thing as OrbitalState - the state vector is always the Cartesian position/velocity. It's just the initialisation that proceeds differently. So the names are misleading, and should be reserved for States whose state vectors conform to that characterisation, which might happen in the future.